src/events/mod.rs

```diff
@@ -0,0 +1,54 @@
+//! Types and traits for use in emitting change notifications
+pub mod defaultsink;
+pub mod filesink;
+use super::events::defaultsink::DefaultEventSink;
+use super::events::filesink::FileEventSink;
+use serde::Serialize;
+
+#[derive(Debug, Serialize)]
+pub enum EventType<T> {
+    InvoiceCreated(T),
+    MissingParcel(T),
+    InvoiceYanked(T),
+    ParcelCreated(T),
+}
+
+#[async_trait::async_trait]
+pub trait EventSink {
+    // Emits an event.
+    async fn raise_event<T>(&self, _: EventType<T>) -> Result<(), Box<dyn std::error::Error>>
+    where
+        T: Serialize + Send + Sync;
+}
+
+// This enum exists to enable the type of event sync to be dynamically selected at runtime depending on arguments passed
+// to the program.
+// It seems like the normal approach of using trait objects does not work as the trait is not object safe as it has a generic type parameter.
```

Comment:
Yep, you are correct here. It is the one glaring weakness of the Trait system. Generally we have not been exposing these publicly and have been putting them in the binary (as anyone can technically write something to implement the trait. Not going to block on this as we are still pre-1.0 and can break this later if needed